pkg/vmconfig/policy/policy_reconciler_util.go

```diff
@@ -0,0 +1,247 @@
+		"imageName", imageName,
+		"imageLabels", imageLabels,
+		"workloadLabels", workloadLabels)
+
+	obj := vspherepolv1.PolicyEvaluation{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: vm.Namespace,
+			Name:      "vm-" + vm.Name,
+		},
+	}
+
+	result, err := controllerutil.CreateOrPatch(ctx, k8sClient, &obj, func() error {
+
+		// Ensure the PolicyEvaluation object is owned by the VM.
+		if err := controllerutil.SetControllerReference(
+			vm,
+			&obj,
+			k8sClient.Scheme()); err != nil {
+
+			return fmt.Errorf(
+				"failed to mark VM %[1]s/%[2]s as owner "+
+					"of PolicyEvaluation %[1]s/%[3]s: %[4]w",
+				vm.Namespace,
+				vm.Name,
+				obj.Name,
+				err)
+		}
+
+		// Ensure explicit policies are evaluated.
+		if vmp := vm.Spec.Policies; len(vmp) == 0 {
+			obj.Spec.Policies = nil
+		} else {
+			obj.Spec.Policies = explicitPolicies
+		}
+
+		// Ensure the image criteria is considered.
+		if imageName == "" && len(imageLabels) == 0 {
+			obj.Spec.Image = nil
+		} else {
+			obj.Spec.Image = &vspherepolv1.PolicyEvaluationImageSpec{
+				Name:   imageName,
+				Labels: imageLabels,
+			}
+		}
+
+		// Ensure the workload criteria is considered.
+		if guestID == "" && guestFamily == "" && len(workloadLabels) == 0 {
+			obj.Spec.Workload = nil
+		} else {
+			obj.Spec.Workload = &vspherepolv1.PolicyEvaluationWorkloadSpec{
+				Labels: workloadLabels,
+				Guest: &vspherepolv1.PolicyEvaluationGuestSpec{
+					GuestID:     guestID,
+					GuestFamily: guestFamily,
+				},
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to create or patch VM %[1]s/%[2]s "+
+				"PolicyEvaluation %[1]s/%[3]s: %[4]w",
+			vm.Namespace,
+			vm.Name,
+			obj.Name,
+			err)
+	}
+
+	waiting := func(reason any) error {
+		logger.Info("Waiting on PolicyEvaluation object",
+			"reason", reason,
+			"generation", obj.Generation,
+			"observedGeneration", obj.Status.ObservedGeneration,
+			"spec", obj.Spec)
+
+		// The object is still being processed.
+		return fmt.Errorf(
+			"VM %[1]s/%[2]s "+
+				"PolicyEvaluation %[1]s/vm-%[2]s still being evaluated: %[3]w",
+			vm.Namespace,
+			vm.Name,
+			ErrPolicyNotReady)
+
+	}
+
+	switch result {
+	case controllerutil.OperationResultCreated,
+		controllerutil.OperationResultUpdated:
+
+		return nil, waiting(result)
+
+	case controllerutil.OperationResultNone:
+
+		if obj.Generation != obj.Status.ObservedGeneration {
+			return nil, waiting("generation")
+		}
+
+		return obj.Status.Policies, nil
+	}
+
+	return nil, fmt.Errorf(
+		"invalid result %[1]s updating to get VM %[2]s/%[3]s "+
+			"PolicyEvaluation %[2]s/vm-%[3]s: %[4]w",
+		result,
+		vm.Namespace,
+		vm.Name,
+		err)
+}
+
+// getTagsFromPolicyEvaluationResults returns the list of vSphere tag UUIDs
+// from one of more PolicyEvaluationResult objects.
+func getTagsFromPolicyEvaluationResults(
+	results ...vspherepolv1.PolicyEvaluationResult) []string {
+
+	var (
+		active    []string
+		activeMap = map[string]struct{}{}
```

Comment:
nit: `sets.Set[string]`, it has builtin func `UnsortedList` to turn into a slice